docs/coding-guidelines.md

```diff
@@ -239,7 +239,36 @@ For example the following are correct:
     }
     ```
 
-Exceptions are allowed when multiple type names coming from different namespaces are available.
+    Exceptions are allowed when multiple type names coming from different namespaces are available.
+
+1. Do not use `as` to cast types
```

Comment:
While I agree that objectively, `as` casting is probably not necessary, this almost feels like one of those things that's too opinionated. I don't mind accepting it, but that was my first instict.

Reply:
Just about everything in the coding guidelines is opinionated!

In cases where nulls are silently ignored, it can be a correctness issue. I'm thinking of something like `var item = CurrentSelection as Package; if (item is not null) { DoSomething(item); }` to handle the case when a button is clicked when no package is selected. If the type is wrong, customers will have selected an item, but nothing happens and the customer doesn't know why, and there are no errors to help us understand.

In cases of `var item = CurrentSelection as Package; var name = item.Name;` then a `NullReferenceException` is much harder to debug from telemetry compared to an `InvalidCastException` that tells use "Type1 cannot be cast to Type2".

So, I don't understand why this guideline would be less appropriate than most of the other opinionated guidelines in this document. Many others are nothing more than stylistic, this one actually reduces debugging complexity, hence increases productivity.

Reply:
This guideline makes sense to me. My ask is that we keep an eye where this guideline has an unintended consequence and revisit the guideline if needed.